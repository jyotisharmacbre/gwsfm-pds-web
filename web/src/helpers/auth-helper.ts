// @ts-ignore
import authentication from '@kdpw/msal-b2c-react';
import jwtDecode from 'jwt-decode';
import { DecodedToken } from '../models/decodedToken';

export const getDisplayName = () => {
  const decoded: DecodedToken = decodeToken();
  return decoded.name;
};

export const getFirstName = () => {
  const decoded: DecodedToken = decodeToken();
  return decoded.given_name;
};

export const getDisplayEmail = () => {
  const decoded: DecodedToken = decodeToken();
  return decoded.email ?
    decoded.email :
    decoded.emails && decoded.emails.length > 0 ?
      decoded.emails[0] :
      '';
};

export const logOut = () => {
  authentication.signOut();
}

function decodeToken() {
  const token = authentication.getAccessToken();
  const decoded: DecodedToken = jwtDecode(token);
  return decoded;
}

