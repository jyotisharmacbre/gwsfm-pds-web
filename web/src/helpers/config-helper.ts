const appConfig = () => {
    const config: any = (window as any).Config;
    if(config && config.REACT_APP_AUTH_INSTANCE==="#{AUTH_INSTANCE}#"){
      return {
        ...process.env
      }
    }
    return {
      ...config
    }
  };
  
  export default appConfig;