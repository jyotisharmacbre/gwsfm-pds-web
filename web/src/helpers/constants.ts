export const selectionButtons = [
  {
    title: 'BUTTON_YES',
    value: true
  },
  {
    title: 'BUTTON_NO',
    value: false
  }
];

export enum Notify {
  none,
  success,
  error
}
