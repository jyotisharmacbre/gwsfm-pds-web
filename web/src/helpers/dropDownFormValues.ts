export const projectStatusData: object[] = [
  { label: '1', value: '1' },
  { label: '2', value: '2' },
  { label: '3', value: '3' }
];

export const engagementData: object[] = [
  { label: 'Type 1', value: '1' },
  { label: 'Type 2', value: '2' },
  { label: 'Type 3', value: '3' }
];

export const typesOfEngagement: object[] = [
  {
    label: '---Please choose an option---',
    value: ''
  },
  {
    label: 'Fixed Price',
    value: 'Fixed Price'
  },
  {
    label: 'Time and Materials',
    value: 'Time and Materials'
  },
  {
    label: 'Risk and Reward',
    value: 'Risk and Reward'
  },
  {
    label: 'Cost Plus',
    value: 'Cost Plus'
  }
];

export const contractType: object[] = [
  {
    label: '---Please choose an option---',
    value: ''
  },
  {
    label: 'CT 1',
    value: 'CT 1'
  },
  {
    label: 'CT 2',
    value: 'CT 2'
  }
];

export const currency: object[] = [
  {
    label: '---Please choose an option---',
    value: ''
  },
  {
    label: 'AED (UAE Dirham)',
    value: 'AED'
  },
  {
    label: 'AFN (Afghani)',
    value: 'AFN'
  },
  {
    label: 'ALL (Lek)',
    value: 'ALL'
  },
  {
    label: 'AMD (Dram)',
    value: 'AMD'
  },
  {
    label: 'ANG (Dutch Antilles Guilder)',
    value: 'ANG'
  },
  {
    label: 'AOA (Kwanza)',
    value: 'AOA'
  },
  {
    label: 'ARS (Peso)',
    value: 'ARS'
  }
];

export const projectStatus: object[] = [
  {
    label: '---Please choose an option---',
    value: ''
  },
  {
    label: 'Initial Customer Inquiry',
    value: 'Initial Customer Inquiry'
  },
  {
    label: 'J&A',
    value: 'J&A'
  },
  {
    label: 'Archived',
    value: 'Archived'
  },
  {
    label: 'Lost',
    value: 'Lost'
  },
  {
    label: 'Order Received',
    value: 'Order Received'
  },
  {
    label: 'CPP',
    value: 'CPP'
  },
  {
    label: 'Complete',
    value: 'Complete'
  }
];

export const assetsWorkedOn: object[] = [
  {
    label: '---Please choose an option---',
    value: ''
  },
  {
    label: 'AS 1',
    value: 'AS 1'
  },
  {
    label: 'AS 2',
    value: 'AS 2'
  }
];

export const enquiryTypeData: object[] = [
  { label: 'Written', value: '1' },
  { label: 'Verbal', value: '2' },
  { label: 'Full tender enquiry with specifications & orgs', value: '3' },
  { label: 'Other', value: '4' }
];
