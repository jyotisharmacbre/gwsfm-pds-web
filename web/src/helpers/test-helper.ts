import checkPropTypes from 'check-prop-types';

export const findByTestAtrr = (component: any, attr: string) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};

export const checkProps = (component: any, expectedProps: any) => {
  const propsErr = checkPropTypes(
    component.propTypes,
    expectedProps,
    'props',
    component.name
  );
  return propsErr;
};
