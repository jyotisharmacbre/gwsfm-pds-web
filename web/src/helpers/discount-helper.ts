import DiscountType from '../enums/DiscountType';
import { restrictMinusAndAllowDecimalForMaxRangeHundred, restrictMinusAndAllowDecimal } from '../helpers/utility-helper';

const normalizeClientDiscount = (discountType: DiscountType, discountValue) => {
    let updatedDiscount;
    if (discountType === DiscountType.Percent) {
        updatedDiscount = restrictMinusAndAllowDecimalForMaxRangeHundred(discountValue);
    }
    else updatedDiscount = restrictMinusAndAllowDecimal(discountValue);
    return updatedDiscount;
}

export default normalizeClientDiscount;