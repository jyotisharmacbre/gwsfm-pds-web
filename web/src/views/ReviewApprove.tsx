import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { match } from 'react-router-dom';
import { History } from 'history';
import CalculationsSummaryTable from '../components/Table/CalculationsSummaryTable';
import PricingSummaryTable from '../components/Table/PricingSummaryTable';
import * as actions from '../store/rootActions';
import { toast } from 'react-toastify';
import { getPropertyName, getFilterElementFromArray } from '../helpers/utility-helper';
import { IState } from '../store/state';
import ProjectSummary from '../components/Forms/ProjectForm/ProjectSummary';
import ProjectOverviewSummary from '../components/Forms/ProjectOverviewForm/ProjectOverviewSummary';
import { IProjectDetail } from '../store/CustomerEnquiryForm/Types/IProjectDetail';
import { ICurrency } from '../store/Lookups/Types/ICurrency';
import Currency from '../store/Lookups/InitialState/Currency';
import { ISubContractorActivity } from '../store/SubContractor/Types/ISubContractorActivity';
import { IPreliminariesComponentDetails } from '../store/Preliminaries/Types/IPreliminariesComponentDetails';
import { IDiscountActivity } from '../store/DiscountForm/Types/IDiscountActivity';
import { ILookup } from '../store/Lookups/Types/ILookup';
import { LookupType } from '../store/Lookups/Types/LookupType';
import { IProjectOverviewDetails } from '../store/ProjectOverviewForm/Types/IProjectOverviewDetails';
import QueryPopup from '../components/Popup/QueryPopup';
import { FormattedMessage } from 'react-intl';
import { IUserServiceData } from '../store/UserService/Types/IUserService';
import ActivityFeedList from '../components/Forms/ProjectOverviewForm/ActivityFeedList';
import EventType from '../enums/EventType';
import IReactIntl from '../Translations/IReactIntl';
import { formatMessage } from '../Translations/connectedIntlProvider';
import { IAdminDefaults } from '../store/Admin/Types/IAdminDefault';
import { ICountry } from '../store/Lookups/Types/ICountry';
import { ICountryHoc, countryHoc } from '../hoc/CountryHoc';
import { insuranceRateHoc, IInsuranceRateHoc } from '../hoc/InsuranceRateHoc';
import { IDynamicBusinessUnits, IDynamicsDivision } from '../store/DynamicsData/Types/IDynamicData';
import { getDisplayEmail } from '../helpers/auth-helper';
import { ProjectSignOffStatus } from '../store/ProjectOverviewForm/Types/ProjectApprovalEnums';
import ProjectStatus from '../enums/ProjectStatus';
import ErrorType from '../enums/ErrorType';
import { IProjectApprovals } from '../store/ProjectOverviewForm/Types/IProjectApprovals';

interface IProps {
	match: match<{ projectId: string }>;
	intl: any;
}

interface IMapStateToProps {
	history: History;
	projectStatus: Array<ILookup>;
	project: IProjectDetail;
	projectOverview: IProjectOverviewDetails;
	subContractorState: Array<ISubContractorActivity>;
	preliminaryState: Array<IPreliminariesComponentDetails>;
	discountState: IDiscountActivity;
	currencies: Array<ICurrency> | null;
	userNamesForEmails: Array<IUserServiceData>;
	initialStateSetForProjectApprovals: boolean;
	lookups: Array<ILookup>;
	adminDefaultValues: Array<IAdminDefaults>;
	countries: Array<ICountry> | null;
	getListOfDivisions: Array<IDynamicsDivision>;
	getListOfBusinessUnit: Array<IDynamicBusinessUnits>;
}

interface IMapDispatchToProps {
	getSubContractor: (projectId: string) => void;
	getPreliminaryDetails: (projectId: string) => void;
	getDiscountData: (projectId: string) => void;
	getAllCurrencies: () => void;
	getProjectDetail: (projectId: string) => void;
	getAdditionalDetails: (projectId: string) => void;
	getProjectStatus: () => void;
	handleGetUserNamesForEmails: (emails: Array<string>) => Array<IUserServiceData>;
	getUserNamesForEmails: (emails: Array<string>) => Array<IUserServiceData>;
	getLookups: () => void;
	getProjectActivities: (projectId: string) => void;
	queryAdd: (projectId: string, formValue: string, event: EventType) => void;
	getProjectParameters: (countryId: number) => void;
	getAllCountries: () => void;
	getDynamicsListOfDivision: () => void;
	getListOfBusinessUnits: () => void;
}

const lookupKeyList: string[] = [
	LookupType.Project_Approval_Range,
	LookupType.Project_Approval_Sign_Off_Status,
	LookupType.Project_Approver_Type
];

const ReviewApprove: React.FC<IProps & IMapStateToProps & IMapDispatchToProps & ICountryHoc & IInsuranceRateHoc> = (
	props
) => {
	const CurrencyObj = new Currency();
	const [currencySymbol, setCurrencySymbol] = useState<string>('');
	const projectId = props.match.params.projectId;
	const [showQueryPopup, setShowQueryPopup] = useState<boolean>(false);
	const [showQueryApproveButton, setShowQueryApproveButton] = useState<boolean>(true);
	const [renderReviewApprove, setRenderReviewApprove] = useState<boolean>(false);

	useEffect(() => {
		window.scrollTo(0, 0);
		props.getAdditionalDetails(projectId);
		props.getAllCurrencies();
		props.getProjectStatus();
		props.getProjectDetail(projectId);
		props.getSubContractor(projectId);
		props.getPreliminaryDetails(projectId);
		props.getDiscountData(projectId);
		props.getLookups();
		props.getProjectActivities(projectId);
		props.getAllCountries();
	}, []);

	useEffect(
		() => {
			if (props.project.currencyId > 0 && props.currencies) {
				setCurrencySymbol(
					getFilterElementFromArray(
						props.currencies,
						getPropertyName(CurrencyObj, (prop) => prop.currencyId),
						props.project.currencyId,
						getPropertyName(CurrencyObj, (prop) => prop.currencySymbol)
					)
				);
			}
		},
		[props.project.currencyId, props.currencies]
	);
	useEffect(
		() => {
			if (props.project.countryId > 0) props.getProjectParameters(props.project.countryId);
		},
		[props.project.countryId]
	);

	// Redirect page to unauthorised page in case of approver status is pending
	useEffect(
		() => {
			if (props.projectOverview.projectId !== '' && props.project.projectId !== '') {
				let email = getDisplayEmail();
				let loggedInApprover = props.projectOverview.projectApprovals.find(x => x.userId?.toLowerCase() === email.toLowerCase());
				if (props.projectOverview.projectApprovals.length === 0 || !loggedInApprover) {
					redirect('Error', ErrorType.unauthorised);
					return;
				}
				handleInvalidApproverAction(loggedInApprover);
			}
		},
		[props.projectOverview, props.project]
	);

	const handleInvalidApproverAction = (loggedInApprover: IProjectApprovals) => {
		let message = "";
		setRenderReviewApprove(true);
		let isReviewRequestApproved = loggedInApprover?.approvalStatus === ProjectSignOffStatus.Approved;
		let isReviewRequestPending = loggedInApprover?.approvalStatus === ProjectSignOffStatus.Pending;
		let isReviewRequestInDraft = loggedInApprover?.approvalStatus === ProjectSignOffStatus.Draft;

		let isProjectInReview = parseInt(props.project.status.toString()) === ProjectStatus.InReview;

		if (isReviewRequestApproved) message = formatMessage('MESSAGE_ALREADY_APPROVED');
		if (isReviewRequestInDraft) message = formatMessage('MESSAGE_NOT_IN_REVIEW');

		if (!(isReviewRequestPending && isProjectInReview)) {
			handleApproveUnauthorizedError(message);
		}
	}

	const redirect = (module: string, errorType: ErrorType) => {
		return props.history.push({
			pathname: `/${module}`,
			state: {
				type: errorType
			}
		});
	};

	/* istanbul ignore next */
	const handleApproval = () => {
		actions.projectApprove(props.match.params.projectId, handleApprovalSuccess, handleApprovalError);
	};
	/* istanbul ignore next */
	const handleApprovalSuccess = (data) => {
		toast.success(formatMessage('MESSAGE_SUCCESSFUL_APPROVED'));
		props.history.push('/');
	};
	/* istanbul ignore next */
	const handleApprovalError = (data) => {
		toast.error(formatMessage('MESSAGE_ERROR'));
	};
	/* istanbul ignore next */
	const handleQuerySuccess = (data) => {
		toast.success(formatMessage('MESSAGE_QUERY_SUCCESS'));
		props.history.push('/');
	};
	/* istanbul ignore next */
	const handleQueryError = (data) => {
		toast.error(formatMessage('MESSAGE_ERROR'));
	};
	/* istanbul ignore next */
	const handleQuerySave = (data: string) => {
		actions.postQuery(props.match.params.projectId, data, handleQuerySuccess, handleQueryError);
	};
	/* istanbul ignore next */
	const actionEditBtn = () => {
		props.history.push(`/Project/${props.match.params.projectId}`);
	}
	/* istanbul ignore next */
	const actionEditBtnOverview = () => {
		props.history.push(`/ProjectOverview/${props.match.params.projectId}`);
	}
	/* istanbul ignore next */
	const handleQueryCancel = () => { setShowQueryPopup(false); }

	const handleApproveUnauthorizedError = (message: string) => {
		if (showQueryApproveButton) {
			toast.warn(message);
			setShowQueryApproveButton(false);
		}
	}

	return (
		<React.Fragment>
			{
				renderReviewApprove &&
				<div className="container-fluid" data-test="review-approve-component">
					{showQueryPopup && (
						<QueryPopup
							intl={props.intl}
							handleConfirm={handleQuerySave}
							handleCancel={handleQueryCancel}
							titleKey={<FormattedMessage id="TITLE_QUERY" />}
							subTitleKey={<FormattedMessage id="SUB_TITLE_QUERY" />}
							contentKey={<FormattedMessage id="PLACEHOLDER_QUERY" />}
						/>
					)}
					<div className="row">
						<div className="col-lg-12">
							<div className="custom-wrap">
								<div className="heading-subtitle">
									<h1>
										<FormattedMessage id="LABEL_REVIEW_APPROVE" />
									</h1>
								</div>
								<ProjectSummary
									oneditclick={actionEditBtn}
									project={props.project}
									lookUpData={props.projectStatus}
									currencySymbol={currencySymbol}
									userNamesForEmails={props.userNamesForEmails}
									handleGetUserNamesForEmails={props.handleGetUserNamesForEmails}
									listOfDivisions={props.getListOfDivisions}
									listOfBusinessUnits={props.getListOfBusinessUnit}
								/>
								<ProjectOverviewSummary
									oneditOverview={actionEditBtnOverview}
									projectOverview={props.projectOverview}
									lookUpData={props.projectStatus}
								/>
								<div className="row">
									<div className="col-xl-9">
										<PricingSummaryTable
											data-test="pricing-summary"
											preliminary={props.preliminaryState}
											subContractor={props.subContractorState}
											discount={props.discountState}
											currencySymbol={currencySymbol}
											insuranceRate={props.insuranceRate}
											countryCode={props.countryCode}
											showDiscount={true}
											showContractor={true}
											showPreliminary={true}
											showInsurance={true}
										/>
										<CalculationsSummaryTable
											data-test="calculation-summary"
											preliminary={props.preliminaryState}
											subContractor={props.subContractorState}
											discount={props.discountState}
											currencySymbol={currencySymbol}
											insuranceRate={props.insuranceRate}
										/>
									</div>
								</div>
								<div className="row">
									<div className="col-xl-12">
										<ActivityFeedList
											data-test="activity-feed-list"
											currencySymbol={currencySymbol}
											handleGetUserNamesForEmails={props.getUserNamesForEmails}
										/>
									</div>
								</div>
								<div className="two-side-btn pt-2">
									<button data-test="btnQuery" type="button" onClick={() => setShowQueryPopup(true)} hidden={!showQueryApproveButton}>
										<FormattedMessage id="BUTTON_QUERY" />
									</button>
									<button data-test="btnApprove" type="button" name="next" onClick={handleApproval} hidden={!showQueryApproveButton}>
										<FormattedMessage id="BUTTON_APPROVE" />
									</button>
								</div>
							</div>
						</div>
					</div>
				</div>
			}
		</React.Fragment>
	);
};

const mapStateToProps = (state: IState) => ({
	projectStatus: state.lookup.projectstatus,
	project: state.project.form,
	subContractorState: state.subContractor.form.activities,
	preliminaryState: state.preliminary.preliminaryDetails,
	discountState: state.discount.form,
	currencies: state.lookup.currencies,
	projectOverview: state.projectOverview.form,
	userNamesForEmails: state.userService.userServiceData,
	lookups: state.lookup.lookups,
	initialStateSetForProjectApprovals: state.projectOverview.initialStateSetForProjectApprovals,
	adminDefaultValues: state.admin.adminDefaultValues,
	countries: state.lookup.countries,
	getListOfDivisions: state.dynamicData.dynamicsListOfDivision,
	getListOfBusinessUnit: state.dynamicData.dynamicsListOfBusinessUnits
});

const mapDispatchToProps = (dispatch) => {
	return {
		getProjectStatus: () => dispatch(actions.getProjectStatus()),
		getSubContractor: (projectId: string) => dispatch(actions.getSubContractor(projectId)),
		getPreliminaryDetails: (projectId: string) => dispatch(actions.getPreliminaryDetails(projectId)),
		getDiscountData: (projectId: string) => dispatch(actions.getDiscountData(projectId)),
		getAllCurrencies: () => dispatch(actions.getAllCurrencies()),
		getProjectDetail: (projectId) => dispatch(actions.getProjectDetail(projectId)),
		getAdditionalDetails: (projectId) => dispatch(actions.getAdditionalDetails(projectId)),
		handleGetUserNamesForEmails: (emails: Array<string>) => dispatch(actions.getUserNamesForEmailsService(emails)),
		getUserNamesForEmails: (emails: Array<string>) => dispatch(actions.getNamesForEmailActivitiesFeed(emails)),
		getLookups: () => dispatch(actions.getLookupsByLookupItems(lookupKeyList)),
		getProjectActivities: (projectId) => dispatch(actions.getProjectActivities(projectId)),
		getProjectParameters: (countryId: number) => dispatch(actions.getProjectParameters(countryId)),
		getAllCountries: () => dispatch(actions.getAllContries()),
		getDynamicsListOfDivision: () => dispatch(actions.getListOfDivision()),
		getListOfBusinessUnits: () => dispatch(actions.getListOfBusinessUnits())
	};
};

export default connect(mapStateToProps, mapDispatchToProps)(countryHoc(insuranceRateHoc(ReviewApprove)));
