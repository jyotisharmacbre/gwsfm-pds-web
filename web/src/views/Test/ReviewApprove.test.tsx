import React from 'react';
import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import ReviewApprove from '../ReviewApprove';
import { findByTestAtrr, checkProps } from '../../helpers/test-helper';
import { IntlProvider } from 'react-intl';
import translations from '../../Translations/translation';

describe('review and approve component test cases', () => {
	let wrapper: any;
	const history = { push: jest.fn() };
	beforeEach(() => {
		wrapper = shallow(<ReviewApprove history={history} match={{ params: { projectId: 1 } }} />);
	});

	it('defines the component', () => {
		expect(wrapper).toBeDefined();
	});

	it('should match the snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});
	it('should renders the component', () => {
		expect(findByTestAtrr(wrapper, 'review-approve-component')).toBeDefined();
	});

	it('should redirect to discount page on previous button click', () => {
		const button = findByTestAtrr(wrapper, 'previous-button');
		button.simulate('click');
		expect(history.push).toHaveBeenLastCalledWith('/Discounts/1');
	});

	it('should renders the pricing summary component', () => {
		expect(findByTestAtrr(wrapper, 'pricing-summary')).toBeDefined();
	});

	it('should renders the calculation summary component', () => {
		expect(findByTestAtrr(wrapper, 'calculation-summary')).toBeDefined();
	});
	
});
