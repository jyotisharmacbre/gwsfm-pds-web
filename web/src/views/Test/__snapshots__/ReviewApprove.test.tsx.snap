// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`review and approve component test cases should match the snapshot 1`] = `ShallowWrapper {}`;
